File: count_and_say.go
package main

import (
	"bytes"
	"fmt"
)

var cache map[int]string

func calculateCache() {
	cache = make(map[int]string)
	cache[1] = "1"
	for index := 2; index <= 30; index++ {
		cache[index] = say(cache[index-1])
	}
}

// 121 => 111211
func say(num string) string {
	var counter [][]int
	var buffer bytes.Buffer

	for _, n := range num {
		nInInt := int(n)
		if len(counter) == 0 {
			counter = append(counter, []int{nInInt, 1})
			continue
		}
		lastWordCounter := counter[len(counter)-1]
		if lastWordCounter[0] == int(n) {
			lastWordCounter[1]++
		} else {
			counter = append(counter, []int{nInInt, 1})
		}
	}

	for _, val := range counter {
		buffer.WriteString(fmt.Sprint(val[1]))
		buffer.WriteString(fmt.Sprintf("%c", val[0]))
	}

	return buffer.String()
}

func count_and_say(n int) string {
	return cache[n]
}

func main() {
	calculateCache()
	fmt.Println("result:", count_and_say(4))
}

File: count_and_say.rb
@cache = {}

def cal_cache
  @cache[1] = "1"
  (2..30).to_a.each do |n|
    prev_num = @cache[n - 1]
    @cache[n] = say(prev_num)
  end
end

# [{1 => 2}, {2 => 1}]
def say(num)
  counter = []
  num.split("").each_with_index do |n, i|
    prev_counter = counter.last
    if prev_counter.nil? or prev_counter.last != n
      counter << [1, n]
    elsif prev_counter[1] == n
      counter[-1][0] += 1
    end
  end
  counter.join
end

def count_and_say(n)
  cal_cache()
  @cache[n]
end

puts count_and_say(5)
